README.md
# RSMP Schemer
Gem for validating RSMP messages against RSMP JSON schemas.

## Background
When communicating via RSMP, the version of core and SXL used is negotiated during connection, and can therefore vary depending on who you communicate with.

To validate against the correct version of the JSON schema, all relevant versions need to be avaiable, so you can 

The actual JSON scheme files are maintained in the repo https://github.com/rsmp-nordic/rsmp_schema. Different version of core and SXLs are mainted in different branches.

Each of these branches are included here as submodules, so you can choose which core and SXL version you want to validate against.

The actual validation is performed with the json_schemer gem.

## Usage
Use ```Schemer#validate``` to validate against a particular schema type and version.

Validating against core and sxl schemas is done separately.
The core schema validates everything that's common to all types of equipment/SXLs.
SXL schemas validate the mesage types specific to a type of equipment.

For example, the core schema will validate that a CommandRequest has all the correct attributes, e.g. that there is a command code. The SXL schema will validate the content of the attributes. e.g. that the command code represents a valid command.

If there are validation errors and error will be returned, pointing to the offendig items.  The errors come from the json_schemer gem.

Nil means there are no errors, ie. the message is valid.

Example, validating an an AggregatedStatusRequest:

```ruby
message = {
  "mType" => "rSMsg",
  "type" => "AggregatedStatusRequest",
  "mId" => "E68A0010-C336-41ac-BD58-5C80A72C7092",
  "cId" => "AB+84001=860SG001"
}

# validing using core 3.1.4 will return an error, because the message
# type AggregatedStatusRequest was not introduced until 3.1.5:
errors = RSMP::Schemer.validate(message, core: '3.1.4')
# => [["/type", "enum", nil]]

# The first item '/type' indicate the field with error
# The second item 'enum' indicate an invalid enum valiue

# validating using core 3.1.5 will succeeed, ie. return no errors:
errors = RSMP::Schemer.validate(message, core: '3.1.5')
# => nil

# validate against several schemas:
errors = RSMP::Schemer.validate(message, core: '3.1.5', tlc: '1.0.15')
# => nil
```



CHANGELOG.md
## [Unreleased]

## [0.1.0] - 2021-03-25
- Initial release

## [0.2.0] - 2021-05-19
- Updated interface

## [0.2.1] - 2021-06-17
- Updated schemas

## [0.2.2] - 2021-06-18
- Option for lenient version string parsing
 


spec/spec_helper.rb
require "bundler/setup"
require "rsmp_schemer"

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/schemer_spec.rb
RSpec.describe RSMP::Schemer do
  it 'has correct schemas' do
    expect(RSMP::Schemer.has_schema?(:bad,'3.1.5')).to be(false)
    expect(RSMP::Schemer.has_schema?(:bad,'1.0.15')).to be(false)

    expect(RSMP::Schemer.has_schema?(:core,'3.1.1')).to be(false)
    expect(RSMP::Schemer.has_schema?(:core,'3.1.2')).to be(true)
    expect(RSMP::Schemer.has_schema?(:core,'3.1.3')).to be(true)
    expect(RSMP::Schemer.has_schema?(:core,'3.1.4')).to be(true)
    expect(RSMP::Schemer.has_schema?(:core,'3.1.5')).to be(true)
    expect(RSMP::Schemer.has_schema?(:core,'3.1.6')).to be(false)

    expect(RSMP::Schemer.has_schema?(:tlc,'1.0.6')).to be(false)
    expect(RSMP::Schemer.has_schema?(:tlc,'1.0.7')).to be(true)
    expect(RSMP::Schemer.has_schema?(:tlc,'1.0.8')).to be(true)
    expect(RSMP::Schemer.has_schema?(:tlc,'1.0.9')).to be(true)
    expect(RSMP::Schemer.has_schema?(:tlc,'1.0.10')).to be(true)
    expect(RSMP::Schemer.has_schema?(:tlc,'1.0.11')).to be(false)
    expect(RSMP::Schemer.has_schema?(:tlc,'1.0.12')).to be(false)
    expect(RSMP::Schemer.has_schema?(:tlc,'1.0.13')).to be(true)
    expect(RSMP::Schemer.has_schema?(:tlc,'1.0.14')).to be(true)
    expect(RSMP::Schemer.has_schema?(:tlc,'1.0.15')).to be(true)
    expect(RSMP::Schemer.has_schema?(:tlc,'1.0.16')).to be(false)
  end

  it 'parses versions strings strictly' do
    expect(RSMP::Schemer.has_schema?(:core,'3.1.5.extra.9.8.7')).to be(false)
    expect(RSMP::Schemer.has_schema?(:tlc,'1.0.15.extra.9.8.7')).to be(false)
  end

  it 'parses versions strings leniently' do
    expect(RSMP::Schemer.has_schema?(:core,'3.1.5.extra.9.8.7', lenient: true)).to be(true)
    expect(RSMP::Schemer.has_schema?(:tlc,'1.0.15.extra.9.8.7', lenient: true)).to be(true)
  end

  it 'raises when schema not found' do
    expect {
      RSMP::Schemer.find_schema!(:bad,'3.1.5')
    }.to raise_error(RSMP::Schemer::UnknownSchemaTypeError)

    expect {
      RSMP::Schemer.find_schema!(:core,'0.0.0')
    }.to raise_error(RSMP::Schemer::UnknownSchemaVersionError)
  end

  it 'returns nil when schema not found' do
    expect(RSMP::Schemer.find_schema(:bad,'3.1.5')).to be_nil
    expect(RSMP::Schemer.find_schema(:bad,'3.1.5')).to be_nil
  end

  it 'finds schemas or return nil' do
    expect(RSMP::Schemer.find_schemas(:core)).to be_a(Hash)
    expect(RSMP::Schemer.find_schemas(:bad)).to be_nil
  end

  it "raise exception when trying to validate against non-existing version" do
    message = {
      "mType" => "rSMsg",
      "type" => "AggregatedStatusRequest",
      "mId" => "E68A0010-C336-41ac-BD58-5C80A72C7092",
      "cId" => "AB+84001=860SG001"
    }
    expect {
      RSMP::Schemer.validate(message, core: '0.0.1')
    }.to raise_error(RSMP::Schemer::UnknownSchemaVersionError)
  end

end


spec/versions_spec.rb
RSpec.describe RSMP::Schemer do

  it "accepts alarm request in core from 3.1.5," do
    message = {
      "mType" => "rSMsg",
      "type" => "AggregatedStatusRequest",
      "mId" => "E68A0010-C336-41ac-BD58-5C80A72C7092",
      "cId" => "AB+84001=860SG001"
    }
    expect(RSMP::Schemer.validate(message, core: '3.1.4')).to be_a(Array)
    expect(RSMP::Schemer.validate(message, core: '3.1.5')).to be_nil
  end

  it "accepts M0104 in tlc from 1.0.15" do
    message = {
      "mType" => "rSMsg",
      "mId" => "4173c2c8-a933-43cb-9425-66d4613731ed",
      "type" => "CommandRequest",
      "siteId" => [
        { "sId" => "RN+SI0001" }
      ],
      "cId" => "O+14439=481WA001",
      "arg" => [
        {
          "cCI" => "M0104",
          "n" => "year",
          "cO" => "setDate",
          "v" => "2021"
        }
      ]
    }
    #expect(RSMP::Schemer.validate(message, tlc: '1.0.14')).to be_a(Array)
    expect(RSMP::Schemer.validate(message, tlc: '1.0.15')).to be_nil
  end

end


lib/rsmp_schemer/schemer.rb
require 'json_schemer'

module RSMP
end

module RSMP::Schemer
  @@schemas = nil

  def self.schemas
    return @@schemas if @@schemas
    schemas_path = File.expand_path( File.join(__dir__,'..','..','schemas') )
    @@schemas = {
      core: {
        '3.1.2' => JSONSchemer.schema( Pathname.new(File.join(schemas_path, 'core_3.1.2/schema/core/rsmp.json')) ),
        '3.1.3' => JSONSchemer.schema( Pathname.new(File.join(schemas_path, 'core_3.1.3/schema/core/rsmp.json')) ),
        '3.1.4' => JSONSchemer.schema( Pathname.new(File.join(schemas_path, 'core_3.1.4/schema/core/rsmp.json')) ),
        '3.1.5' => JSONSchemer.schema( Pathname.new(File.join(schemas_path, 'core_3.1.5/schema/core/rsmp.json')) )
      },
      # note that tlc 1.0.11 and 1.0.12 does not exist (unreleased drafts)
      tlc: {
        '1.0.7' => JSONSchemer.schema( Pathname.new(File.join(schemas_path, 'tlc_1.0.7/schema/tlc/sxl.json')) ),
        '1.0.8' => JSONSchemer.schema( Pathname.new(File.join(schemas_path, 'tlc_1.0.8/schema/tlc/sxl.json')) ),
        '1.0.9' => JSONSchemer.schema( Pathname.new(File.join(schemas_path, 'tlc_1.0.9/schema/tlc/sxl.json')) ),
        '1.0.10' => JSONSchemer.schema( Pathname.new(File.join(schemas_path, 'tlc_1.0.10/schema/tlc/sxl.json')) ),
        '1.0.13' => JSONSchemer.schema( Pathname.new(File.join(schemas_path, 'tlc_1.0.13/schema/tlc/sxl.json')) ),
        '1.0.14' => JSONSchemer.schema( Pathname.new(File.join(schemas_path, 'tlc_1.0.14/schema/tlc/sxl.json')) ),
        '1.0.15' => JSONSchemer.schema( Pathname.new(File.join(schemas_path, 'tlc_1.0.15/schema/tlc/sxl.json')) )
      }
    }
  end

  def self.validate_using_schema message, schema
    raise ArgumentError.new("message missing") unless message
    raise ArgumentError.new("schema missing") unless schema
    unless schema.valid? message
      schema.validate(message).map do |item|
        [item['data_pointer'],item['type'],item['details']]
      end
    else
      []
    end
  end

  def self.find_schemas type
    raise ArgumentError.new("type missing") unless type
    schemas[type.to_sym]
  end

  def self.find_schemas! type
    schemas = find_schemas type
    raise UnknownSchemaTypeError.new("Unknown schema type #{type}") unless schemas
    schemas
  end

  def self.find_schema type, version, options={}
    raise ArgumentError.new("version missing") unless version
    version = sanitize_version version if options[:lenient]
    if version
      schemas = find_schemas type
      return schemas[version] if schemas
    end
    nil
  end

  def self.sanitize_version version
    matched = /^\d+\.\d+\.\d+/.match version
    matched.to_s if matched
  end

  def self.find_schema! type, version, options={}
    schema = find_schema type, version, options
    raise ArgumentError.new("version missing") unless version
    version = sanitize_version version if options[:lenient]
    if version
      schemas = find_schemas! type
      schema = schemas[version]
      return schema if schema
    end
    raise UnknownSchemaVersionError.new("Unknown schema version #{type} #{version}")
  end

  def self.has_schema? type, version, options={}
    find_schema(type,version, options) != nil
  end

  def self.validate message, schemas, options={}
    raise ArgumentError.new("message missing") unless message
    raise ArgumentError.new("schemas missing") unless schemas
    raise ArgumentError.new("schemas must be a Hash") unless schemas.is_a?(Hash)
    raise ArgumentError.new("schemas cannot be empty") unless schemas.any?
    errors = schemas.flat_map do |type, version|
      schema = find_schema! type, version, options
      validate_using_schema(message, schema)
    end
    return nil if errors.empty?
    errors
  end

end


examples/validate.rb
require_relative '../lib/rsmp_schemer.rb'

# What schemas are avaible?
puts "Available schemas: #{RSMP::Schemer.schema_versions}"
puts

['3.1.4','3.1.5'].each do |version|
  puts "Has core schema #{version}: #{RSMP::Schemer.has_schema?(:core,version)}"
end
['1.0.7','1.0.15'].each do |version|
  puts "Has tlc schema #{version}: #{RSMP::Schemer.has_schema?(:tlc,version)}"
end
puts


#validate valid message against core and tlc schemas
message = {
  "mType" => "rSMsg",
  "mId" => "4173c2c8-a933-43cb-9425-66d4613731ed",
  "type" => "CommandRequest",
  "siteId" => [
    { "sId" => "RN+SI0001" }
  ],
  "cId" => "O+14439=481WA001",
  "arg" => [
    {
      "cCI" => "M0002",
      "n" => "timeplan",
      "cO" => "setPlan",
      "v" => "1"
    }
  ]
}
puts "First let's validate a legit message"
puts "core: #{RSMP::Schemer.validate(message, :core,  '3.1.5') || 'ok'}"
puts "tlc: #{RSMP::Schemer.validate(message, :tlc, '1.0.15') || 'ok'}"
puts


#validate bad message against core and tlc schemas
message = {
  "mType" => "rSMsg",
  "mId" => "4173c2c8-a933-43cb-9425-66d4613731ed",
  "type" => "CommandRequest",
  "siteId" => [
    { "sId" => "RN+SI0001" }
  ],
  "arg" => [
    {
      "cCI" => "M3002",
      "n" => "timeplan",
      "cO" => "setPlan",
      "v" => "1"
    }
  ]
}
puts "Now let's look at a shady message"
puts "core: #{RSMP::Schemer.validate(message, :core,  '3.1.5') || 'ok'}"
puts "tlc: #{RSMP::Schemer.validate(message, :tlc, '1.0.15') || 'ok'}"


